### DISCUSSION TITLE
INAV Compass discrepancy, interference, solutions (thanks to Joshua Bardwell)

### DISCUSSION THREAD

**Discussion by Colechr56** [iNavFlight/inav#10299] [Show and tell]
I had big problems with compass errors, showing random values, at all orientations. Joshua Bardwell has a YouTube video of this similar situation, and his solutions. He found that the XT60 power connector was too close to the GPS/Compass module. He also foil wrapped the leads from the flight controller to the GPS module, but felt this was less significant. So distance, and screening.
For me, I replaced the GPS/compass cables with a screened cable, which I further attached the screen to the power ground line.
I have a BN-880 GPS, SpeedyBeeF405WingApp, FrSky X8R, 5 servos (for on board testing,) and a TX800 VTX (connected with supplied cable and unscreened.
The biggest change appears to be replacing the supplied GPS multi-strand cables, for a grounded screened, multicore cable, soldered to the original Molex GPS 6 pin connector, and the F405 connector. I did put some copper foil over the power leads as they came up to the FC, and a patch over the XT60 connector. The VTX is on the opposite corner of the testing box/board.
The result is that the GPS locks to 3D, quite quickly, when looking through the patio windows, the Compass shows stable compass values, which within a few degrees of expected at the major points of the compass, eg nominally North +-1deg, East +-101deg, South +-164deg, and West +-266deg. There was very little drifting, and leaving the board static for 5 mins, the value didn't change. Now to repeat the test at the pond. But this is significantly better. See the attached photo, of the test board/box.
![20240813_092730](https://github.com/user-attachments/assets/fcfbe7ee-ca01-4f76-9d08-b9414307488f)

**Comment by sensei-hacker** [comment]
Thanks for sharing that.

There are two kinds of fields. (Or two components of an electromagnetic field.) Radiative aka electric fields radiate from the source out into space. An example is an antenna intentionally creates a radiative electric field to send data out.  The other other type of field is an inductive/magnetic field. As you may have noticed when you put a magnet on your refrigerator, the magnetic field stays close to the magnet, so the magnet sticks only when it gets very close to the refrigerator. The magnetic field doesn't radiate out into space.

Adding an electrical shield , electrically grounded, can be useful when dealing with an electric field. Interference from an electric field shows up as missed readings (the I2C 'errors' count). A camera can be sensitive to interference from an electric field, so an electrically shielded cable can make sense on the *camera*.

The magnetometer (compass) measures the _magnetic_ field.
An electric shield is not useful for protecting from _magnetic_ field interference. You can test this yourself by putting a piece of foil in between two magnets.  You will see the magnets still stick together. The electric "shielding" doesn't stop the magnetic field at all. With a **magnet**ometer, it's **magnetic** fields that are the problem, so shielding from an electric field isn't generally useful.

The main power wires form an electro magnet, and that's the biggest source of magnetic field interference. Also buzzers like the Vifly produce magnetic fields. How can we deal with those? Remember the inductive magnetic field doesn't radiate - it stays close to the source. So the key to dealing with magnetic interference is **DISTANCE**. Specifically, doubling the distance reduces the interference by 8X.

Again you can see that easily yourself with a refrigerator magnet - the magnetic attraction gets much, much weaker as the distance between the magnet to the fridge increases.

The key is to have as much distance as possible between the magnetometer and the main power wires and the buzzer. Ideally, putting them at opposite ends of a quad.

At the same time, we'd like to have our sensitive GPS antenna as far as possible from the powerful VTX antenna, and also the telemetry antenna. So we want the main power wires, VTX antenna, and telemetry antenna at one end of the quad; the GPS/magnetometer at the other end.

To avoid shadowing in forward flight, the VTX and telemetry antennas should be at the rear. So the GPS/mag can go on the other end - at the front. Above the camera is often a good spot.


In summary:
* The mag needs distance from the power wires and buzzer
* The GPS needs distance from the VTX and telemetry antenna
* Electrically shielding doesn't help the magnetometer
* So put the GPS/mag on the front, away from the power wires and antennas

**Reply by anti-vaxxer** [reply]
> So the key to dealing with magnetic interference is DISTANCE. Specifically, doubling the distance reduces the interference by 8X.

hey @sensei-hacker is it really so because I reckon according to inverse square law that doubling distance decreases intensity by four times?

**Reply by sensei-hacker** [reply]
The *electric (radiative)* field is a square law because is spreads out as a sphere. Surface area of a sphere is pi x r². The energy is spread over that sphere.

The *magnetic* field doesn't radiate; it's attached to the source. Which is why permanent magnets are a thing - the magnetic field hangs around without needing a constant power source because it's not sending power out into the world.

Since the field is attached to the source, the total drops off with distance AND it's spread over the sphere. So pi x r² x r.

**Reply by Bookmdanno** [reply]
The info presented here about electric and magnetic fields is mostly INcorrect.  If you want to understand more about the subject, these are some good places to start - Khan Academy for beginner-friendly explanations, MIT OpenCourseWare for in-depth lectures from a top university, YouTube lectures by Walter Lewin (MIT) for engaging and thorough explanations, and Stanford University's online electromagnetism courses for advanced content.  All offer a variety of topics within electromagnetic theory.
